pkg/block/mdfetcher.go

```diff
@@ -0,0 +1,103 @@
+package block
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	mu             *sync.RWMutex
+
+	concurrencyLimit int
+	logger           log.Logger
+	bkt              objstore.BucketReader
+}
+
+func NewMetadataFetcher(logger log.Logger, concurrencyLimit int, bkt objstore.BucketReader) *MetadataFetcher {
+	return &MetadataFetcher{
+		logger:           logger,
+		concurrencyLimit: concurrencyLimit,
+		mu:               &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadata is retrieved successfully.
+func (f *MetadataFetcher) Fetch(ctx context.Context) (map[ulid.ULID]*metadata.Meta, error) {
+	blockIDs := make(chan ulid.ULID, f.concurrencyLimit)
+	defer close(blockIDs)
+	remoteBlockMDs := make(map[ulid.ULID]*metadata.Meta, len(f.cachedBlockMDs))
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < f.concurrencyLimit; i++ {
+		go func() {
+			for blockID := range blockIDs {
+				md, err := f.fetchBlockMDIfNotCached(ctx, wg, blockID)
+				if err != nil {
+					level.Debug(f.logger).Log("msg", "failed to fetch block metadata", "block", blockID, "error", err)
+				}
+
+				f.mu.Lock()
+				remoteBlockMDs[blockID] = md
+				f.mu.Unlock()
+			}
+		}()
+	}
+
+	if err := f.bkt.Iter(ctx, "", func(name string) error {
+		wg.Add(1)
+
+		id, isBlock := IsBlockDir(name)
+		if !isBlock {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case blockIDs <- id:
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	wg.Wait()
+
+	f.mu.Lock()
```

Comment:
Why this lock? To make fetcher thread-safe? I think we need single flight here TBH with some expiry even (: Can you add `TODO(bwplotka): Add single flight with expiry if needed`

Also, I am not sure if we can really reuse this lock IMO...  I would rather have:
either 
* two locks, one for concurrency here and second for potential concurrent use of `Fetch` 
* remove this lock. I don't think we do currently fetch concurrently so we could remove this and comment that it's not thread safe

I would say let's add `TODO`